cumulus/client/cirrus-executor/src/bundle_processor.rs

```diff
@@ -472,35 +472,50 @@ where
 			.extract_receipts(&BlockId::Hash(primary_hash), extrinsics.clone())?;
 
 		for signed_receipt in signed_receipts.iter() {
+			let secondary_hash = signed_receipt.execution_receipt.secondary_hash;
 			match crate::aux_schema::load_execution_receipt::<
 				_,
 				Block::Hash,
 				NumberFor<PBlock>,
 				PBlock::Hash,
-			>(&*self.client, signed_receipt.execution_receipt.secondary_hash)?
+			>(&*self.client, secondary_hash)?
 			{
 				Some(local_receipt) => {
 					if let Some(trace_mismatch_index) =
 						find_trace_mismatch(&local_receipt, &signed_receipt.execution_receipt)
 					{
-						crate::aux_schema::write_bad_receipt::<_, PBlock>(
+						let trace_mismatch_index = trace_mismatch_index
+							.try_into()
+							.expect("Trace mismatch index must fit into u32; qed");
+
+						crate::aux_schema::write_bad_receipt::<_, PBlock, _>(
 							&*self.client,
 							signed_receipt.execution_receipt.primary_number,
 							signed_receipt.hash(),
-							trace_mismatch_index
-								.try_into()
-								.expect("Trace mismatch index must fit into u32; qed"),
+							(trace_mismatch_index, secondary_hash),
 						)?;
 					}
 				},
 				None => {
+					let block_number: BlockNumber = signed_receipt
+						.execution_receipt
+						.primary_number
+						.try_into()
+						.unwrap_or_else(|_| panic!("Primary number must fit into u32; qed"));
+
+					let block_hash = self.client.hash(block_number.into())?.ok_or_else(|| {
```

Comment:
I'm not sure it canonicalizes anything by default, though we have added finality on the consensus side. I have seen MUCH older blocks being imported just fine if there wasn't finality, so I think we need to have TODOs in every place that is potentially problematic why we can recall it so we can fix it later.

Reply:
Updated TODO, but it's not super easy to describe these edge cases, please take another look at the comment description, hope it can be understood by everyone.